From ffbf7cfc002141d39ebb0a0a9a667b2fce400b49 Mon Sep 17 00:00:00 2001
From: Zhikui Ren <zhikui.ren@intel.com>
Date: Wed, 17 Jan 2024 10:54:21 -0800
Subject: [PATCH] Remove exception during restartControlLoops

During pid-control init phase, an exception is thrown when fan pid class
has invalid number of outputs. This can result in no fan speed control
for the platform even for the good fan pid class. Exception is also
thrown when dbus call to get sensor property fails and this can happen
when sensor hosting service is busy.

Replace these exceptions with error message and continue to allow the
pid control service to run with other configuration that it can parse
correctly.

Tested:
Code compiles and EB Build is provided to test team to run.
CPU thermtrip was not reproduced.

Signed-off-by: Zhikui Ren <zhikui.ren@intel.com>
---
 dbus/dbusconfiguration.cpp | 19 +++++++++++++++----
 sensors/builder.cpp        | 25 +++++++++++++++----------
 sensors/builder.hpp        |  2 +-
 sensors/sensor.hpp         | 24 ++++++++++++++++++++++++
 4 files changed, 55 insertions(+), 15 deletions(-)

diff --git a/dbus/dbusconfiguration.cpp b/dbus/dbusconfiguration.cpp
index 4b7ae38..10b27f0 100644
--- a/dbus/dbusconfiguration.cpp
+++ b/dbus/dbusconfiguration.cpp
@@ -505,8 +505,8 @@ bool init(sdbusplus::bus_t& bus, boost::asio::steady_timer& timer,
         catch (const sdbusplus::exception_t&)
         {
             // this shouldn't happen, probably means daemon crashed
-            throw std::runtime_error("Error getting managed objects from " +
-                                     owner.first);
+            std::cerr << "Error getting managed objects from " << owner.first << "\n";
+            continue;
         }
 
         for (auto& pathPair : configuration)
@@ -774,8 +774,19 @@ bool init(sdbusplus::bus_t& bus, boost::asio::steady_timer& timer,
                     }
                     else
                     {
-                        throw std::runtime_error(
-                            "fan PID has invalid number of Outputs");
+                        /* This can be due to some of other sensor services that host fan sensors are not working.
+                         * Log error and continue, this allows the fan control to work with other good fan pid class
+                         */
+                        std::cerr << "fan PID has invalid number of Outputs: input size " << inputSensorInterfaces.size() << " output size " << outputSensorInterfaces.size() << "\n";
+                        for (const SensorInterfaceType& inputSensorInterface : inputSensorInterfaces)
+                        {
+                            std::cerr << "fan PID has invalid number of Outputs: input " << inputSensorInterface.first << " : " << inputSensorInterface.second << "\n";
+                        }
+                        for (const SensorInterfaceType& outputSensorInterface : outputSensorInterfaces)
+                        {
+                            std::cerr << "fan PID has invalid number of Outputs: output " << outputSensorInterface.first << " : " << outputSensorInterface.second << "\n";
+                        }
+                        continue;
                     }
                     std::string fanSensorName;
                     std::string pwmPath;
diff --git a/sensors/builder.cpp b/sensors/builder.cpp
index 7c335c6..35cf6ce 100644
--- a/sensors/builder.cpp
+++ b/sensors/builder.cpp
@@ -43,20 +43,21 @@ namespace pid_control
 static constexpr bool deferSignals = true;
 
 SensorManager
-    buildSensors(const std::map<std::string, conf::SensorConfig>& config,
+    buildSensors(std::map<std::string, conf::SensorConfig>& config,
                  sdbusplus::bus_t& passive, sdbusplus::bus_t& host)
 {
     SensorManager mgmr(passive, host);
     auto& hostSensorBus = mgmr.getHostBus();
     auto& passiveListeningBus = mgmr.getPassiveBus();
 
-    for (const auto& it : config)
+    for (auto it = config.begin(); it != config.end(); )
     {
+        auto current = it++;
         std::unique_ptr<ReadInterface> ri;
         std::unique_ptr<WriteInterface> wi;
 
-        std::string name = it.first;
-        const conf::SensorConfig* info = &it.second;
+        std::string name = current->first;
+        const conf::SensorConfig* info = &current->second;
 
         std::cerr << "Sensor: " << name << " " << info->type << " ";
         std::cerr << info->readPath << " " << info->writePath << "\n";
@@ -96,9 +97,11 @@ SensorManager
                 }
                 if (ri == nullptr)
                 {
-                    throw SensorBuildException(
-                        "Failed to create dbus passive sensor: " + name +
-                        " of type: " + info->type);
+                    std::cerr << " Failed to create dbus passive sensor :" << name << " of type: " << info->type << "\n";
+                    config.erase(current);
+                    auto sensor = std::make_unique<FailedSensor>(name, info->timeout);
+                    mgmr.addSensor(info->type, name, std::move(sensor));
+                    continue;
                 }
                 break;
             case IOInterfaceType::EXTERNAL:
@@ -147,9 +150,11 @@ SensorManager
 
                     if (wi == nullptr)
                     {
-                        throw SensorBuildException(
-                            "Unable to create write dbus interface for path: " +
-                            info->writePath);
+                        std::cerr <<  "Unable to create write dbus interface for path: " << info->writePath << "\n";
+                        config.erase(current);
+                        auto sensor = std::make_unique<FailedSensor>(name, info->timeout);
+                        mgmr.addSensor(info->type, name, std::move(sensor));
+                        continue;
                     }
 
                     break;
diff --git a/sensors/builder.hpp b/sensors/builder.hpp
index 14eabad..d4bb818 100644
--- a/sensors/builder.hpp
+++ b/sensors/builder.hpp
@@ -16,7 +16,7 @@ namespace pid_control
  * Build the sensors and associate them with a SensorManager.
  */
 SensorManager
-    buildSensors(const std::map<std::string, conf::SensorConfig>& config,
+    buildSensors(std::map<std::string, conf::SensorConfig>& config,
                  sdbusplus::bus_t& passive, sdbusplus::bus_t& host);
 
 } // namespace pid_control
diff --git a/sensors/sensor.hpp b/sensors/sensor.hpp
index a135c50..49ae39b 100644
--- a/sensors/sensor.hpp
+++ b/sensors/sensor.hpp
@@ -66,4 +66,28 @@ class Sensor
     int64_t _timeout;
 };
 
+class FailedSensor : public Sensor
+{
+  public:
+    FailedSensor(const std::string& name, int64_t timeout) :
+        Sensor(name, timeout)
+    {}
+
+    virtual ReadReturn read(void)
+    {
+        return ReadReturn{};
+    };
+
+    virtual void write(double value)
+    {
+        (void) value;
+    };
+
+    virtual bool getFailed(void)
+    {
+        return true;
+    };
+
+};
+
 } // namespace pid_control
-- 
2.25.1

